Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm

```diff
@@ -0,0 +1,589 @@
+
+    // We take the time before m_firstVideoFrameAudioTime is set so that the first frame will always apepar to have a timestamp of 0 but with a longer duration.
+    auto sampleTime = m_firstVideoFrameAudioTime ? nextVideoFrameTime() : resumeVideoTime();
+
+    if (!m_firstVideoFrameAudioTime)
+        m_firstVideoFrameAudioTime = sampleTime;
+
+    ASSERT(m_lastEnqueuedRawVideoFrame <= sampleTime);
+    ASSERT(m_lastEnqueuedRawVideoFrame >= m_lastReceivedCompressedVideoFrame);
+    m_lastEnqueuedRawVideoFrame = sampleTime;
+    m_pendingVideoFrames.append({ Ref { frame }, sampleTime });
+    LOG(MediaStream, "appendVideoFrame:enqueuing raw video frame:%f queue:%zu first:%f last:%f", sampleTime.value(), m_pendingVideoFrames.size(), m_pendingVideoFrames.first().second.value(), m_pendingVideoFrames.last().second.value());
+
+    encodePendingVideoFrames();
+}
+
+Seconds MediaRecorderPrivateWriterWebM::nextVideoFrameTime() const
+{
+    ASSERT(!m_hasAudio || m_firstVideoFrameAudioTime);
+    auto now = MonotonicTime::now();
+    auto frameTime = now - m_resumedVideoTime;
+    LOG(MediaStream, "nextVideoFrameTime: frameTime:%f nextVideoFrameTime:%f", frameTime.value(), (frameTime + *m_firstVideoFrameAudioTime).value());
+    frameTime = frameTime + *m_firstVideoFrameAudioTime;
+    // Take the max value from m_lastAudioSampleTime to handle the occasional wall clock time drift and ensure muxed frames times
+    // are always monotonically increasing.
+    return std::max(frameTime, Seconds { m_lastAudioSampleTime.toDouble() });
+}
+
+Seconds MediaRecorderPrivateWriterWebM::resumeVideoTime() const
+{
+    return m_hasAudio ? Seconds { m_resumedAudioTime.toDouble() } : m_currentVideoDuration;
+}
+
+Seconds MediaRecorderPrivateWriterWebM::sampleTime(const RetainPtr<CMSampleBufferRef>& sample) const
+{
+    return Seconds { PAL::CMTimeGetSeconds(PAL::CMSampleBufferGetPresentationTimeStamp(sample.get())) };
+}
+
+void MediaRecorderPrivateWriterWebM::encodePendingVideoFrames()
+{
+    assertIsCurrent(MainThreadDispatcher::singleton());
+
+    if (m_pendingVideoFrames.isEmpty() || !m_videoEncoder)
+        return;
+
+    m_pendingVideoEncode++;
+
+    Vector<Ref<VideoEncoder::EncodePromise>> promises { m_pendingVideoFrames.size() , [&](auto) {
+        auto frame = m_pendingVideoFrames.takeFirst();
+        bool needVideoKeyframe = false;
+        if (frame.second - m_lastVideoKeyframeTime >= m_maxClusterDuration) {
+            needVideoKeyframe = true;
+            m_lastVideoKeyframeTime = frame.second;
+        }
+        LOG(MediaStream, "encodePendingVideoFrames:encoding video frame:%f", frame.second.value());
+        return m_videoEncoder->encode({ WTFMove(frame.first), frame.second.microsecondsAs<int64_t>(), { } }, needVideoKeyframe);
+    } };
+
+    VideoEncoder::EncodePromise::all(WTFMove(promises))->whenSettled(MainThreadDispatcher::singleton(), [weakThis = ThreadSafeWeakPtr { *this }, this](auto&&) {
```

Comment:
Ce serait plus logique de faire m_pendingVideoEncode++ pour chaque appel m_videoEncoder->encode, et m_pendingVideoEncode-- dans chaque promise retournée.
Plus besoin de VideoEncoder::EncodePromise::all.

Reply:
That would need to dispatch twice as many tasks. The use of promise::all here is slightly more efficient